packages/openapi3/test/shared-routes.test.ts

```diff
@@ -0,0 +1,291 @@
+import { expectDiagnostics } from "@typespec/compiler/testing";
+import { deepStrictEqual } from "assert";
+import { diagnoseOpenApiFor, openApiFor } from "./test-host.js";
+
+describe("openapi3: shared routes", () => {
+  it("emits warning for routes containing query parameters", async () => {
+    const diagnostics = await diagnoseOpenApiFor(
+      `
+      @service({title: "My Service"})
+      namespace Foo {
+        model Resource {
+          id: string;
+        }
+
+        @route("/sharedroutes/resources?filter=resourceGroup")
+        op listByResourceGroup(...Resource): Resource[];
+
+        @route("/sharedroutes/resources?filter=subscription")
+        op listBySubscription(...Resource): Resource[];
+      }
+      `
+    );
+    expectDiagnostics(diagnostics, [
+      {
+        code: "@typespec/openapi3/path-query",
+        message: "OpenAPI does not allow paths containing a query string.",
+      },
+      {
+        code: "@typespec/openapi3/path-query",
+        message: "OpenAPI does not allow paths containing a query string.",
+      },
+    ]);
+  });
+
+  it("model shared routes that differ by query params as a union of optional params", async () => {
+    const results = await openApiFor(
+      `
+      @service({title: "My Service"})
+      namespace Foo {
+        model Resource {
+          id: string;
+        }
+
+        @route("/sharedroutes/resources", { shared: true })
+        op listByResourceGroup(...Resource, @query resourceGroup: string, @query foo: string): Resource[];
+
+        @route("/sharedroutes/resources", { shared: true })
+        op listBySubscription(...Resource, @query subscription: string, @query foo: string): Resource[];
+      }
+      `
+    );
+    const params = results.paths["/sharedroutes/resources"].post.parameters as {
+      name: string;
+      required: boolean;
+    }[];
+    deepStrictEqual(params, [
+      {
+        in: "query",
+        name: "resourceGroup",
+        required: false,
+        schema: {
+          type: "string",
+        },
+      },
+      {
+        in: "query",
+        name: "foo",
+        required: true,
+        schema: {
+          type: "string",
+        },
+      },
+      {
+        in: "query",
+        name: "subscription",
+        required: false,
+        schema: {
+          type: "string",
+        },
+      },
+    ]);
+  });
+
+  it("model shared routes that differ by query param values as an enum", async () => {
+    const results = await openApiFor(
+      `
+      @service({title: "My Service"})
+      namespace Foo {
+        model Resource {
+          id: string;
+        }
+
+        @route("/sharedroutes/resources", { shared: true })
+        op listByResourceGroup(...Resource, @query filter: "resourceGroup"): Resource[];
```

Comment:
What. happens if you have a different return type?

Reply:
I already added those tests per @markcowl's request and am working on making them pass.